torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5778,7 +5778,15 @@ def test_broadcast_object_list(self):
                 else [None for _ in COLLECTIVES_OBJECT_TEST_LIST]
             )
 
-            # Single object test
+            # Single object test with device specified
+            if backend == "nccl":
```

Comment:
@rohan-varma [requested](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599812) a test for non-current CUDA device for both `nccl` and `gloo` backends, but this test only covers `nccl` backend.

I think the `if backend == "nccl":` can probably be removed, and the test should still pass since `gloo` works with GPU as well as mentioned by @rohan-varma [in this comment](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599220).

Reply:
Good point, reverted the if statement. 

dumb Q, not quite familiar with this test. The way I run it is 

for gloo
- BACKEND='gloo' WORLD_SIZE=4 with-proxy python test/distributed/test_distributed_fork.py TestDistBackendWithFork.test_broadcast_object_list --v

for nccl

BACKEND='nccl' WORLD_SIZE=4 with-proxy python test/distributed/test_distributed_fork.py TestDistBackendWithFork.test_broadcast_object_list --v


is that enough?

Reply:
I am also not fully familiar with the distributed testing, but since `backend` is read from the environment variable, I am fairly sure that the way you tested is correct.

I would just wait to see how the CI tests go since I am not sure if the test is ever run with Gloo backend with no GPUs available. In that case, I think that this [line](https://github.com/pytorch/pytorch/blob/1eb70690dcd975bc6ae006bda2406126eb4ac3a8/torch/testing/_internal/distributed/distributed_test.py#L5786) in your test will error due to referencing a non-existing CUDA device.

If I were you, I would duplicate the test you added so that there are two versions: (1) using `torch.device("cpu")` as the specified `device` and (2) using `next_rank` as the specified `device`. In particular, only run (2) if `torch.cuda.device_count() == dist.get_world_size()`, so you know for sure that the line mentioned above will not error.

Reply:
Great idea, I assume the tests executed twice in two different env variable backend=gloo vs backend=nccl. 
Just setting the device conditionally should be enough
device = torch.device('cpu') if backend == 'gloo' else torch.device(next_rank)

Reply:
IIUC, this does not test Gloo with GPU then right? (only NCCL with GPU and Gloo with CPU)

Quoting from @rohan-varma [here](https://github.com/pytorch/pytorch/pull/61305#discussion_r666599220):

> Regarding Gloo backend, Gloo is also frequently used with GPU

I was thinking that if a main part of the motivation for adding this `device` argument is so that users can specify GPU for Gloo backend, then we should test it.

Reply:
Got it, duplicates the test a few times to make sure we can test gloo -- cpu and gpu (next_rank) as well as nccl - gpu(next_rank). How does current version sound like?

Reply:
+1, It would be great if GPU device could be supported in Gloo and a UT for it.

Reply:
done, added three tests here